controllers/route/route_certificaterequest_controller.go

```diff
@@ -0,0 +1,343 @@
+	var instance *routev1.Route
+	err := c.GetClient().Get(ctx, req.NamespacedName, instance)
+	if err != nil {
+		return c.ManageError(ctx, instance, err)
+	}
+	log := c.Log.WithValues("route", req.NamespacedName)
+
+	// The isCertManagerRoute predicate means we should only receive routes we care about.
+	// but the TLS block could be missing / unset so set a default.
+	if instance.Spec.TLS == nil {
+		instance.Spec.TLS = &routev1.TLSConfig{
+			// If no TLS Config was specified, use Edge. This controller
+			// is not responsible for destinationCA, so reencrypt doesn't
+			// make sense. If a user knows they want reencrypt, they can set it themselves.
+			Termination: routev1.TLSTerminationEdge,
+		}
+	}
+
+	// Check if the certificate is up-to-date and not expired.
+	upToDate, renewalTime := func(route *routev1.Route) (bool, time.Time) {
+		pk, err := cmutilpki.DecodePrivateKeyBytes([]byte(route.Spec.TLS.Key))
+		if err != nil {
+			log.V(5).Info("Route has missing or invalid private key", "err", err)
+			return false, time.Time{}
+		}
+		certs, err := cmutilpki.DecodeX509CertificateChainBytes([]byte(route.Spec.TLS.Certificate))
+		if err != nil {
+			log.V(5).Info("Route has missing or invalid certificate", "err", err)
+			return false, time.Time{}
+		}
+		// if no error, certs[0] is the leaf cert (or self-signed cert)
+		matches, err := cmutilpki.PublicKeyMatchesCertificate(pk, certs[0])
+		if err != nil {
+			log.V(5).Info("Couldn't check if Route private key matches Route certificate", "err", err)
+			return false, time.Time{}
+		}
+		if !matches {
+			log.V(5).Info("Route private key doesn't match route cert")
+			return false, time.Time{}
+		}
+		// if we get here, we have a matching TLS cert/key pair.
+		// See if the cert will expire soon.
+		start := certs[0].NotBefore
+		end := certs[0].NotAfter
+		lifetime := end.Sub(start)
+		remaining := end.Sub(time.Now())
+		if remaining <= lifetime/3 {
+			return false, time.Time{}
+		}
+		return true, start.Add(2 * lifetime / 3)
+	}(instance)
+
+	if upToDate {
+		return c.ManageSuccessWithRequeue(ctx, instance, renewalTime.Sub(time.Now()))
+	}
+
+	// If we get here we need to fire off a certificate request. First step is to generate a new
+	// private key.
+	nextPrivateKeyPEM, found := instance.GetAnnotations()[cmapi.IsNextPrivateKeySecretLabelKey]
+	if !found {
+		// Generate a new private key, then re-reconcile.
+		// TODO: make key type and size configurable with an annotation
+		pk, err := cmutilpki.GenerateRSAPrivateKey(cmutilpki.MinRSAKeySize)
+		if err != nil {
+			return c.ManageError(ctx, instance, err)
+		}
+		pkPEM, err := cmutilpki.EncodePrivateKey(pk, cmapi.PKCS1)
+		if err != nil {
+			return c.ManageError(ctx, instance, err)
+		}
+		instance.GetAnnotations()[cmapi.IsNextPrivateKeySecretLabelKey] = string(pkPEM)
+		err = c.GetClient().Update(ctx, instance)
+		if err != nil {
+			return c.ManageError(ctx, instance, err)
+		}
+		return c.ManageSuccess(ctx, instance)
```

Comment:
I think `ManageError` causes these log lines:

```
$  go run . -zap-devel=false -zap-stacktrace-level=panic -zap-log-level=debug
...
{"level":"debug","ts":1639386824.9775126,"logger":"util.api","msg":"object is not ConditionsAware, not setting status"}
{"level":"debug","ts":1639386825.2267056,"logger":"util.api","msg":"object is not ConditionsAware, not setting status"}
{"level":"debug","ts":1639386825.2765381,"logger":"util.api","msg":"object is not ConditionsAware, not setting status"}
{"level":"debug","ts":1639386825.5309203,"logger":"util.api","msg":"object is not ConditionsAware, not setting status"}
{"level":"debug","ts":1639386825.5738258,"logger":"util.api","msg":"object is not ConditionsAware, not setting status"}
{"level":"debug","ts":1639386825.8363264,"logger":"util.api","msg":"object is not ConditionsAware, not setting status"}
{"level":"debug","ts":1639386825.877797,"logger":"util.api","msg":"object is not ConditionsAware, not setting status"}
```

Source: https://github.com/redhat-cop/operator-utils/blob/b22e79c476d3c075953718c8adbece4a0ae1c9f5/pkg/util/reconciler.go#L335

I have no idea how to get around this though :(

Reply:
This isn't an error as such, but I could just not use their ManageError / ManageSuccess functions